rfc/004-priority-based-mempool.md

```diff
@@ -0,0 +1,205 @@
+    // Keep a cache of already-seen txs.
+    // This reduces the pressure on the proxyApp.
+    cache     txCache
+    preCheck  PreCheckFunc
+    postCheck PostCheckFunc
+
+    config *cfg.MempoolConfig
+
+    // Exclusive mutex for Update method to prevent concurrent execution of
+    // CheckTx or ReapMaxBytesMaxGas(ReapMaxTxs) methods.
+    updateMtx sync.RWMutex
+
+    wal          *auto.AutoFile // a log of mempool txs
+    proxyAppConn proxy.AppConnMempool
+
+    logger  log.Logger
+    metrics *Metrics
+}
+
+type mempoolImpl interface {
+    Size() int
+
+    addTx(*mempoolTx, uint64)
+    removeTx(types.Tx) bool // return whether corresponding element is removed or not
+    updateRecheckCursor()
+    reapMaxTxs(int) types.Txs
+    reapMaxBytesMaxGas(int64, int64) types.Txs // based on priority
+    recheckTxs(proxy.AppConnMempool)
+    isRecheckCursorNil() bool
+    getRecheckCursorTx() *mempoolTx
+    getMempoolTx(types.Tx) *mempoolTx
+    deleteAll()
+    // ...and more
+}
+```
+
+Such that mempool's shared code will live under `basemempool` (callback handling, WAL, etc.) while different `mempoolImpl` only needs to implement required methods on transaction addition / removal / iteration etc. The proposed interface is [implemented in this code base](https://github.com/QuarkChain/tendermintx/blob/master/mempool/mempool.go).
+
+### Phase 2: Implement mempool based on different data structures
+
+We will have more discussions on what data structure fits the requirements best. Max-heap or self-balance trees are potential candidates. If concurrent read / write is needed, lock-free data structures (such as [an AVL variant](http://ddana.cswp.cs.technion.ac.il/wp-content/uploads/sites/19/2015/12/logicalorderingavl.pdf)) are also worth considering.
+
+### Concurrency Issues
+
+> Dev and others also expressed some concerns about concurrency patters - “My concern is primarily that we currently have the downside that you can't keep receiving / broadcasting txs while updating / rechecking.“ (see [this issue](https://github.com/tendermint/tendermint/issues/2147) and [this one](https://github.com/tendermint/tendermint/issues/2484))
+
+A smaller tree can be used to receive transactions and broadcast them to other peers. This tree will be merged into the main one once rechecking is done.
+
+> “I'm not sure that that concurrent AVL tree is sufficient for determinism...” Jae
+
+When fee is always zero (for those applications who don’t use fee), it may be possible to issue sequence numbers and use them to sort transactions.
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Positive
+
+- ABCI blockchains will have more customizability on how transactions are included in blocks
+
+### Negative
+
+- At early phases, using priority-based mempools could face performance degradations
+- Mempool implementations may have increased complexities
+
+### Neutral
+
+- Mempool code could undergo a non-trivial refactoring
+
+## Appendix A. Other Projects' Memory Pools
```

Comment:
Would be helpful to provide some summarization here of how both the Libra and GoEthereum mempools work for context. Likely we can borrow quite a bit from them.

Reply:
libra mempool maintains a hashmap of accounts (each account contains a list of transactions, sorted by sequence number). Btree is used for `PriorityIndex` (global index of transactions sorted by priority).

Reply:
Right. Let's add that to the text. Maybe we need ResponseCheckTx to include an "address" field so txs can be grouped according to their "sender" (whatever that is from tendermint's perspective lol) which will help manage some of the complexity here. Iirc, ethereum does something similar.

This is a case where being agnostic to txs is going to be challenging. Either we leak more through the abci boundary or we give the app more control here.

Reply:
that's exactly the problem we faced when trying to simulate the "nonce" system from ethereum. implementing something like that could be useful but I think it's beyond the scope of this RFC and deserves its own.

will update the texts on how it's done in libra and eth.